### ISSUE TITLE
IO Type factory based on methods from the core type?

### ISSUE THREAD

**Issue by zepumph** [phetsims/tandem#188]
Based on patterns that are getting established for IO Types in natural-selection and MultipleParticleModelIO, most of the logic in an IO Type is actually getting defined in the core type, and so IO Types often are mostly boilerplate:

```js

class BunnyIO extends ObjectIO {

  /**
   * Serializes a Bunny instance.
   * @param {Bunny} bunny
   * @returns {Object}
   * @public
   * @override
   */
  static toStateObject( bunny ) {
    validate( bunny, this.validator );
    return bunny.toStateObject();
  }

  /**
   * Deserializes a Bunny instance.
   * @param {Object} stateObject
   * @returns {Object}
   * @public
   * @override
   */
  static fromStateObject( stateObject ) {
    return Bunny.fromStateObject( stateObject );
  }

  /**
   * Creates the args to BunnyGroup.createNextElement that creates Bunny instances.
   * @param state
   * @returns {Object[]}
   * @public
   * @override
   */
  static stateToArgsForConstructor( state ) {
    return Bunny.stateToArgsForConstructor( state );
  }

  /**
   * Restores Bunny state after instantiation.
   * @param {Bunny} bunny
   * @param {Object} state
   * @public
   * @override
   */
  static applyState( bunny, state ) {
    validate( bunny, this.validator );
    bunny.applyState( state );
  }
}
```

In those cases, it could be nice to have an IO Type factory that can just wrap the core type.

**Comment by samreid** [comment]
It seems like we should coordinate this work with the other IO type work proposed in 

https://github.com/phetsims/phet-io/issues/1657#issuecomment-646735445
https://github.com/phetsims/phet-io/issues/1646
https://github.com/phetsims/phet-io/issues/1643
https://github.com/phetsims/phet-io/issues/1657#issuecomment-646735445

Do we need an IO type label?

**Comment by samreid** [comment]
In https://github.com/phetsims/phet-io/issues/1683#issuecomment-646229929 @pixelzoom said:

> >  In other issues we have decided that, by convention, many functions of the IO Type will just forward a method over to the core type/instance, I.E. BunnyIO.toStateObject calls Bunny.prototyp.toStateObject.
> 
> The reason for this convention is that serialization/deserialization typically involves accessing private fields of the core type.  If the IO Type accesses those fields, it's violating the core type's API.   If the IO Type delegates to the core type, then it's not a problem.

**Comment by samreid** [comment]
Here's a patch that introduces a new function `createIOType` and also demonstrates its use for `BunnyIO`.  The usage site looks like this:

```js
createIOType( Bunny, 'BunnyIO', 'IO Type for Bunny', {} )
```

All validation & forwarding is set up automatically.  If there are any methods, they can be supplied in the options.

Additionally, if we declare it within the during the declaration for Bunny.js, we can have direct type checks without using the namespace and without any circular module loading problems:

`Bunny.BunnyIO = createIOType( Bunny, 'BunnyIO', 'IO Type for Bunny', {} )`

I'd like to get feedback from @pixelzoom and/or @zepumph before moving more in this direction, but it looks promising to me.

**Comment by pixelzoom** [comment]
I couldn't apply the patch to natural-selection master HEAD for some reason, lots of merge conflicts.  But I inspected the code and I like this direction.  A few suggestions:

* Having to import `createIOType` is inconvenient. Can `createIOType` be a static method of `PhetioObject`?  Is there any reason to create an IO Type for something that's not a `PhetioObject`? 
* Document `createIOType`
* Make `options` an optional parameter of `createIOType`.  It's currently required, and call sites are providing `{}`.
* Move `phetioDocumentation` parameter of `createIOType` to `options`, with default`'IO Type of ${typeName}'`, so we don't have to repeat this boilerplate doc over and over.

**Comment by samreid** [comment]
I addressed the review comments and committed PhetioObject.createIOType.  One recommendation I did not complete is:

> * Move `phetioDocumentation` parameter to `options`, with default`'IO Type for ${typeName}'`, so we don't have to repeat "IO Type for Blah" over and over.

The `typeName` is the IO type name, like `BunnyIO` so this wouldn't work as a default unless we strip off the `IO`.  Also, at a recent meeting we wanted to investigate leveraging core type JSDoc for this documentation, but we haven't investigated that part yet in https://github.com/phetsims/phet-io/issues/1691

This was far enough along that I was comfortable committing it, but one deficiency that will need to be addressed, is that several times in the PhET-iO engine it checks the existence of a method to decide what to do.  In this commit, every IO type provides an implementation for all methods, like `toStateObject`.  To address this, we could:
* Only add these methods dynamically if they are defined in the core type (somehow)
* Add another way of checking whether to invoke these methods.
* Invoke the methods and use their return value accordingly.

**Comment by pixelzoom** [comment]
This is looking good.

Can we move `documentation` to options and default it to `''`?  Since `IOType.documentation` is optional, I don't think it should be a required parameter to `createIOType`. 

I don't fully understand the implications of the remaining phetioEngine "deficiency".  Is  there a performance impact? 
Does it need to be addressed before this can be used in production code?

**Comment by samreid** [comment]
Thanks, I moved `documentation` to options and defaulted to `''`.

> I don't fully understand the implications of the remaining phetioEngine "deficiency". Is there a performance impact?
Does it need to be addressed before this can be used in production code?

There are several places where we check the existence of a method to determine how to proceed.  For instance, phetioCommandProcessor.js

```js
    // Serialize the result if possible.
    return methodSignature.returnType.toStateObject ?
           methodSignature.returnType.toStateObject( result ) :
           result;
```

With `PhetioObject.createIOType` as it is, `toStateObject` will always be defined.  There are a few other methods and call sites like this.  It must be addressed before production usage.

I think we should move toward a pattern where `toStateObject` is always defined, but some implementations will just return `result`.

**Comment by zepumph** [comment]
Beginning of review:
* If we really want this to be the end all, we need to support:
  * Super types (@samreid hopes for no supertypes
  * parametric types
* We need to make sure that we aren't sacrificing the value of IO types as public schema and documentation in exchange for an easier time coding them up. @samreid and I discussed a few places where this is currently happening. If we don't support inheritance, how will methods or constructor logic be transferred instead of duplicated (like for DerivedPropertyIO taking the `link` method from `PropertyIO`. We also mentioned how transferring all logic onto the core type is great in many cases, but doesn't address the potential need for the IO type hierarchy to deviate from the core type hierarchy.

* In regards to the `toStateObject ?toStateObject( result ): null` check. We discussed and think that is out of date. ObjectIO now defines all these methods, and that chunk can now be written like:

<details>

```
    const args = command.args.map( ( arg, index ) => {
      const TypeIO = methodSignature.parameterTypes[ index ];
      assert && assert( !( typeof arg === 'object' && TypeIO.wrapForPhetioCommandProcessor ), 'unsupported arg TypeIO' );
      if ( typeof arg === 'string' && TypeIO.wrapForPhetioCommandProcessor ) {
        return this.wrapFunction( targetWindow, arg, TypeIO, request );
      }
      else {
        assert && assert( TypeIO.supportsDataTypeSerialization );
        return TypeIO.fromStateObject( arg );
      }
    } );

    // Invoke the method.  If it throws an error, it will be caught in handleSingleInvoke and returned to the client
    // See https://github.com/phetsims/phet-io/issues/941
    const result = methodSignature.implementation.apply( wrapper, args );

    // Serialize the result if possible.
    return methodSignature.returnType.toStateObject( result );
  }

```

</details>

**Comment by pixelzoom** [comment]
> Super types (@samreid hopes for no supertypes

I can see how we could very easily have had a need for an IO Type hierarchy in Natural Selection.  Part of the model class hierarchy in NS looks like this:

```
PhetioObject
  Organism
    Bunny
    Wolf
```

Bunny and Wolf are dynamic elements. I feel like I lucked out that everything stateful in Organism was a Property. If I'd had even 1 non-instrumented field, I would have needed an IO Type hierarchy.  (And no, I don't think that BunnyIO and WolfIO should have been reaching down into Organism to duplicate the saving of its state. That is an OOP anti pattern.)

So I agree that we need to support IO Type hierarchy. But could `createIOType` be useful for the (typical?) case where no such hierarchy is needed?  It certain could be used to replace 100% of sim-specific IO Types in NS.   And if you need an IO Type hierarchy, or parameterized types, then use the more verbose approach of defining an IO type in its own .js file.

> .... doesn't address the potential need for the IO type hierarchy to deviate from the core type hierarchy.

This deviation is something that I think is generally to be avoided, really undesirable.

**Comment by zepumph** [comment]
>But could createIOType be useful for the (typical?) case where no such hierarchy is needed?

That is the best place to start in my opinion. I would rather risk having multiple patterns for IO Types than to over commit to this potentially limiting pattern.

Back to you @samreid fo the above review comments.

**Comment by samreid** [comment]
Marking as blocks publication until the preceding commit has been reviewed and/or passed CT.

**Comment by samreid** [comment]
This patch demonstrates using this pattern for `BunnyIO`

<details>

```diff

Index: js/common/model/BunnyIO.js
===================================================================
--- js/common/model/BunnyIO.js	(revision 1c22312b90c1c2c7562848ae4c7f4e9c9f7247d9)
+++ js/common/model/BunnyIO.js	(revision 1c22312b90c1c2c7562848ae4c7f4e9c9f7247d9)
@@ -1,58 +0,0 @@
-// Copyright 2020, University of Colorado Boulder
-
-/**
- * BunnyIO is the IO Type for Bunny. It delegates most of its implementation to Bunny.
- *
- * @author Chris Malley (PixelZoom, Inc.)
- */
-
-import validate from '../../../../axon/js/validate.js';
-import ObjectIO from '../../../../tandem/js/types/ObjectIO.js';
-import naturalSelection from '../../naturalSelection.js';
-import Bunny from './Bunny.js';
-
-class BunnyIO extends ObjectIO {
-
-  /**
-   * Serializes a Bunny instance.
-   * @param {Bunny} bunny
-   * @returns {Object}
-   * @public
-   * @override
-   */
-  static toStateObject( bunny ) {
-    validate( bunny, this.validator );
-    return bunny.toStateObject();
-  }
-
-  /**
-   * Creates the args to BunnyGroup.createNextElement that creates Bunny instances.
-   * @param state
-   * @returns {Object[]}
-   * @public
-   * @override
-   */
-  static stateToArgsForConstructor( state ) {
-    return Bunny.stateToArgsForConstructor( state );
-  }
-
-  /**
-   * Restores Bunny state after instantiation.
-   * @param {Bunny} bunny
-   * @param {Object} state
-   * @public
-   * @override
-   */
-  static applyState( bunny, state ) {
-    validate( bunny, this.validator );
-    bunny.applyState( state );
-  }
-}
-
-BunnyIO.documentation = 'IO Type for Bunny';
-BunnyIO.validator = { isValidValue: value => value instanceof Bunny };
-BunnyIO.typeName = 'BunnyIO';
-ObjectIO.validateSubtype( BunnyIO );
-
-naturalSelection.register( 'BunnyIO', BunnyIO );
-export default BunnyIO;
\ No newline at end of file
Index: js/common/model/BunnyArrayIO.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/BunnyArrayIO.js	(revision 1c22312b90c1c2c7562848ae4c7f4e9c9f7247d9)
+++ js/common/model/BunnyArrayIO.js	(date 1598845652939)
@@ -10,10 +10,10 @@
 import ObjectIO from '../../../../tandem/js/types/ObjectIO.js';
 import ReferenceIO from '../../../../tandem/js/types/ReferenceIO.js';
 import naturalSelection from '../../naturalSelection.js';
+import Bunny from './Bunny.js';
 import BunnyArray from './BunnyArray.js';
-import BunnyIO from './BunnyIO.js';
 
-class BunnyArrayIO extends ObservableArrayIO( ReferenceIO( BunnyIO ) ) {}
+class BunnyArrayIO extends ObservableArrayIO( ReferenceIO( Bunny.BunnyIO ) ) {}
 
 BunnyArrayIO.documentation = 'IO Type for BunnyArray';
 BunnyArrayIO.validator = { isValidValue: value => value instanceof BunnyArray };
Index: js/common/model/BunnyGroup.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/BunnyGroup.js	(revision 1c22312b90c1c2c7562848ae4c7f4e9c9f7247d9)
+++ js/common/model/BunnyGroup.js	(date 1598845652945)
@@ -15,7 +15,6 @@
 import Tandem from '../../../../tandem/js/Tandem.js';
 import naturalSelection from '../../naturalSelection.js';
 import Bunny from './Bunny.js';
-import BunnyIO from './BunnyIO.js';
 import EnvironmentModelViewTransform from './EnvironmentModelViewTransform.js';
 import GenePool from './GenePool.js';
 
@@ -37,7 +36,7 @@
 
       // phet-io
       tandem: Tandem.REQUIRED,
-      phetioType: PhetioGroupIO( BunnyIO ),
+      phetioType: PhetioGroupIO( Bunny.BunnyIO ),
       phetioDocumentation: 'manages dynamic PhET-iO elements of type Bunny, including live and dead bunnies'
     }, options );
 
Index: js/common/model/Bunny.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/Bunny.js	(revision 1c22312b90c1c2c7562848ae4c7f4e9c9f7247d9)
+++ js/common/model/Bunny.js	(date 1598845886963)
@@ -15,14 +15,15 @@
 import merge from '../../../../phet-core/js/merge.js';
 import required from '../../../../phet-core/js/required.js';
 import AssertUtils from '../../../../phetcommon/js/AssertUtils.js';
+import PhetioObject from '../../../../tandem/js/PhetioObject.js';
 import Tandem from '../../../../tandem/js/Tandem.js';
 import BooleanIO from '../../../../tandem/js/types/BooleanIO.js';
 import NullableIO from '../../../../tandem/js/types/NullableIO.js';
 import NumberIO from '../../../../tandem/js/types/NumberIO.js';
+import ObjectIO from '../../../../tandem/js/types/ObjectIO.js';
 import ReferenceIO from '../../../../tandem/js/types/ReferenceIO.js';
 import naturalSelection from '../../naturalSelection.js';
 import NaturalSelectionUtils from '../NaturalSelectionUtils.js';
-import BunnyIO from './BunnyIO.js';
 import EnvironmentModelViewTransform from './EnvironmentModelViewTransform.js';
 import GenePool from './GenePool.js';
 import Genotype from './Genotype.js';
@@ -67,7 +68,7 @@
 
       // phet-io
       tandem: Tandem.REQUIRED,
-      phetioType: BunnyIO,
+      phetioType: Bunny.BunnyIO,
       phetioDynamicElement: true
     }, options );
 
@@ -329,8 +330,8 @@
       generation: NumberIO.toStateObject( this.generation ),
       age: NumberIO.toStateObject( this.age ),
       isAlive: BooleanIO.toStateObject( this.isAlive ),
-      father: NullableIO( ReferenceIO( BunnyIO ) ).toStateObject( this.father ),
-      mother: NullableIO( ReferenceIO( BunnyIO ) ).toStateObject( this.mother ),
+      father: NullableIO( ReferenceIO( Bunny.BunnyIO ) ).toStateObject( this.father ),
+      mother: NullableIO( ReferenceIO( Bunny.BunnyIO ) ).toStateObject( this.mother ),
       genotype: GenotypeIO.toStateObject( this.genotype ),
       phenotype: PhenotypeIO.toStateObject( this.phenotype ),
 
@@ -372,8 +373,8 @@
     this.generation = required( NumberIO.fromStateObject( stateObject.generation ) );
     this.age = required( NumberIO.fromStateObject( stateObject.age ) );
     this.isAlive = required( BooleanIO.fromStateObject( stateObject.isAlive ) );
-    this.father = required( NullableIO( ReferenceIO( BunnyIO ) ).fromStateObject( stateObject.father ) );
-    this.mother = required( NullableIO( ReferenceIO( BunnyIO ) ).fromStateObject( stateObject.mother ) );
+    this.father = required( NullableIO( ReferenceIO( Bunny.BunnyIO ) ).fromStateObject( stateObject.father ) );
+    this.mother = required( NullableIO( ReferenceIO( Bunny.BunnyIO ) ).fromStateObject( stateObject.mother ) );
     this.genotype.applyState( stateObject.genotype );
     this.phenotype.applyState( stateObject.phenotype );
 
@@ -436,5 +437,7 @@
   return new Vector3( dx, dy, dz );
 }
 
+Bunny.BunnyIO = PhetioObject.createIOType( Bunny, 'BunnyIO', ObjectIO );
+
 naturalSelection.register( 'Bunny', Bunny );
 export default Bunny;
\ No newline at end of file
Index: js/common/model/SelectedBunnyProperty.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/SelectedBunnyProperty.js	(revision 1c22312b90c1c2c7562848ae4c7f4e9c9f7247d9)
+++ js/common/model/SelectedBunnyProperty.js	(date 1598845652943)
@@ -14,7 +14,7 @@
 import NullableIO from '../../../../tandem/js/types/NullableIO.js';
 import ReferenceIO from '../../../../tandem/js/types/ReferenceIO.js';
 import naturalSelection from '../../naturalSelection.js';
-import BunnyIO from './BunnyIO.js';
+import Bunny from './Bunny.js';
 
 class SelectedBunnyProperty extends Property {
 
@@ -30,7 +30,7 @@
 
       // phet-io
       tandem: Tandem.REQUIRED,
-      phetioType: PropertyIO( NullableIO( ReferenceIO( BunnyIO ) ) ),
+      phetioType: PropertyIO( NullableIO( ReferenceIO( Bunny.BunnyIO ) ) ),
       phetioDocumentation: 'the selected bunny, null if no bunny is selected'
     }, options );
 
```

</details>

**Comment by samreid** [comment]
In the commit, I added support for inheritance using the `class extends ParentType` pattern which reminds me of our new mixin pattern described in https://github.com/phetsims/phet-info/issues/143.

**Comment by samreid** [comment]
I leveraged this pattern to create parametric types for ActionIO and EmitterIO and the following tests work:

* Axon unit tests (including EmitterIOTests in phet-io brand)
* Studio works
* Mirror inputs works
* In natural selection studio, this listener works:

```js
phetio.phetioClient.invoke('naturalSelection.introScreen.view.resetAllButton.firedEmitter','addListener',[()=>{console.log('hello');}]);
```

This eliminates a lot of boilerplate, fixes module load order and turns it into a more declarative style (through options).  I'm not sure what else to test before committing.

**Comment by samreid** [comment]
The proposed pattern is working well, but I wondered if we are painting ourselves into a corner by not using traditional `class ... extends` hierarchies, like `class ChargedParticleIO extends ModelElementIO`.  But to bind to the core type requires something mixin-like: `class ModelElementIO extends BoundObjectIO(ModelElement)` which may not be better than the given proposal.  I'd like to think about it and discuss it, and perhaps be able to change to this pattern if we like it better, before there are too many calls to `PhetioObject.createIOType`.

**Comment by zepumph** [comment]
This pattern is looking really nice! @samreid took me through it today. We think it is ready to be applied to all IO types. 
I think that it makes the most sense to name this the same for each type, like `IOType`, and then for parameteric types something like `createIOType()`.

One small change that will limit the boilerplate is to move the parentType into the options, and to generally infer the parent IO type from the parent of the core type, something like `ChargedParticle.prototype.constructor.IOType`. @samreid wasn't sure if that is available in a built sim.

**Comment by pixelzoom** [comment]
> ... We think it is ready to be applied to all IO types. ...

Tracking for natural-selection in https://github.com/phetsims/natural-selection/issues/218.  

> One small change that will limit the boilerplate is to move the parentType into the options, ...

Just so I understand... Are you proposing to change the function signature from this:

````PhetioObject.createIOType = ( CoreType, typeName, ParentType = ObjectIO, options )```

to this?

```PhetioObject.createIOType = ( CoreType, typeName, options )```

If so, yes I agree. Especially since `ParentType` has a default value specified in the current signature.

Let me know when this has been done, so that I can proceed with https://github.com/phetsims/natural-selection/issues/218.

**Comment by pixelzoom** [comment]
In the above commit, I rename param `ParentType` to `ParentIOType`.

In `createIOType`:

```js
    // {string} e.g., "Animal that has a genotype (genetic blueprint) and a phenotype (appearance)."
    documentation: `IO type for ${typeName.substring( 0, typeName.length - 2 )}`,
```

I'm wondering if this could be implemented in a way that makes it more obvious that we're stripping off "IO".

**Comment by pixelzoom** [comment]
... or if you can't find a way to make it more obvious that we're stripping off "IO", then make it so it won't break if the suffix is ever changed, and so you're not duplication 'IO' in the assertion.  Something like (untested):

```js
const IO_TYPE_SUFFIX = 'IO';

PhetioObject.createIOType = ( CoreType, typeName, ParentIOType = ObjectIO, options ) => {
  assert && assert( typeName.endsWith( IO_TYPE_SUFFIX ) || typeName.includes( `${IO_TYPE_SUFFIX }<` ), `IO type name must end with ${IO_TYPE_SUFFIX}` );
...
    documentation: `IO type for ${typeName.substring( 0, typeName.length - IO_TYPE_SUFFIX.length )}`,

**Comment by pixelzoom** [comment]
> And are there other place in PhET-iO codebase where 'IO' is hardcoded?

In ObjectIO, this is not good. The literal 'IO' is duplicated 3x in one line!

```js
    assert && assert( splitOnParameters.indexOf( 'IO' ) === splitOnParameters.length - 'IO'.length, 'type name must end with IO' );
```

It should be:

```js
const IO_TYPE_SUFFIX = 'IO';
...
    assert && assert( splitOnParameters.indexOf( IO_TYPE_SUFFIX ) === splitOnParameters.length - IO_TYPE_SUFFIX.length, `type name must end with ${IO_TYPE_SUFFIX}` );
```

**Comment by pixelzoom** [comment]
Same problem in PhetioElementView.js, duplicated twice here:

```js
    assert && assert( _.endsWith( baseTypeName, 'IO' ), 'IO typename should end in "IO"' );
```

**Comment by pixelzoom** [comment]
> And are there other place in PhET-iO codebase where 'IO' is hardcoded?

Yeah, this is another problem. See https://github.com/phetsims/tandem/issues/205.

**Comment by pixelzoom** [comment]
Note that in https://github.com/phetsims/tandem/issues/206, `createIOType` was moved from PhetioObject to ObjectIO.  In case you're wondering where it went...

**Comment by pixelzoom** [comment]
I started converting NS in https://github.com/phetsims/natural-selection/issues/218, starting with BunnyIO.  I deleted BunnyIO, then added this in Bunny.js:

`Bunny.BunnyIO = ObjectIO.createIOType( Bunny, 'BunnyTypeIO' );`

BunnyGroup.js now has this:

`      phetioType: PhetioGroupIO( Bunny.BunnyIO ),`

And here's what I see in Studio for `naturalSelection.introScreen.model.bunnyCollection.bunnyGroup.archetype`:

<img width="548" alt="screenshot_544" src="https://user-images.githubusercontent.com/3046552/92070968-c6739000-ed6a-11ea-962d-c4e048889e1b.png">

"BunnyTypeIO" is not at all what I expected for IO Type.  Is something wrong here?

**Comment by pixelzoom** [comment]
Oops, I see my bug. Nevermind. Deleting previous comment.

**Comment by pixelzoom** [comment]
I'm attempting to convert NS to use `createIOType` in https://github.com/phetsims/natural-selection/issues/218.

Does this approach not work with `ReferenceIO` subtypes, or maybe parameterized types?  For example in NS, AlleleIO is:

`class AlleleIO extends ReferenceIO( ObjectIO ) {}`

I attempted to replace in Allele.js with:

```js
Allele.AlleleIO = ObjectIO.createIOType( Allele, 'AlleleIO', {
  parentIOType: ReferenceIO( ObjectIO )
} );
```

Then GenePair.js becomes:

```js
  toStateObject() {
    return {
      gene: GeneIO.toStateObject( this.gene ),
148   fatherAllele: Allele.AlleleIO.toStateObject( this.fatherAllele ),
      motherAllele: Allele.AlleleIO.toStateObject( this.motherAllele )
    };
  }
```

And line 148 fails with:

> Uncaught TypeError: phetioObject.toStateObject is not a function

**Comment by pixelzoom** [comment]
More trouble converting NS...  

The IO Type created by `createIOType` requires the Core Type to have 4 methods:

```
toStateObject
static fromStateObject
static stateToArgsForConstructor
applyState
```

All of my Core Types are missing one or more of those required methods. For example, GenePair is missing `fromStateObject`, and fails immediately in the State wrapper.   

I'm not sure how this is working for ChargedParticle, because it's missing `fromStateObject` and `stateToArgsForConstructor`.  Maybe those methods aren't called for that IO Types. But in any case, `createIOType` is creating a broken class.

Questions:

(1) I didn't discover this until I ran the State Wrapper. Can `createIOType` please validate that the Core Type has all of the required methods, so this can be detected more immediately?

(2) Should `PhetioObject` provide default implementations of the methods required for the Core Type?

**Comment by pixelzoom** [comment]
Re the problem in the previous issue... 

ObjectIO was providing the default implementation of the required methods. Because (for example) `BunnyIO extends ObjectIO`, BunnyIO could implement only the methods that it needs to override. That is no longer the case with `createIOType`.  The core type is now responsible for implementing all required methods, and gets no help from ObjectIO.   So we need to somehow get those default implementations from ObjectIO back into the implementation.

**Comment by pixelzoom** [comment]
>  We need to somehow get those default implementations from ObjectIO back into the implementation.

Maybe something like this in the class generated by `createIOType`?

```diff
     static fromStateObject( stateObject ) {
-      return coreType.fromStateObject( stateObject );
+      if ( coreType.hasOwnProperty( 'fromStateObject' ) ) {
+        return coreType.fromStateObject( stateObject );
+      }
+      else {
+        return options.parentIOType.fromStateObject( stateObject );
+      }
     }
```

This still has a bit of a smell to it. We're delegating to the Core Type by overriding the `fromStateObject` method. Then Core Type is delegating right back to the parent IO Type.

**Comment by pixelzoom** [comment]
Another problem that is really painful for me to point out…  Parameter names and type expressions in the IO Type generated by `createIOType` should match those of ObjectIO.  For some reason _every one of the parameters has either a different name, a different type expression, or both!!!_

Compare:

```
// ObjectIO
   * @param {*} o
   * @returns {*}
   * @public
   */
  static toStateObject( o ) {

// createIOType
     * @param {PhetioObject} phetioObject
     * @returns {Object}
     * @public
     * @override
     */
    static toStateObject( phetioObject ) {
```

```
// ObjectIO
   * @param {*} o - whatever was returned from the toStateObject method
   * @returns {*} - the deserialized instance of the same type that the toStateObject was provided as a parameter
   * @public
   */
  static fromStateObject( o ) {

// createIOType
    // @public
static fromStateObject( stateObject ) {
```

```
// ObjectIO
 * @param {Object} stateObject - from a corresponding`toStateObject` method
 * @returns {Array.<*>} - the array of arguments to be passed to creation functions like PhetioGroup.create() or PhetioCapsule.getElement().
 * @public
 */
static stateToArgsForConstructor( stateObject ) {


// createIOType
     * @param state
     * @returns {Object[]}
     * @public
     * @override
     */
    static stateToArgsForConstructor( state ) {
```

```
// ObjectIO
 * @param {PhetioObject} o
 * @param {Object} value - result from toStateObject
 * @public
 */
static applyState( o, value ) { }


// createIOType
     * @param {PhetioObject} phetioObject
     * @param {Object} state
     * @public
     * @override
     */
    static applyState( phetioObject, state ) {
```

**Comment by pixelzoom** [comment]
So to summarize, `createIOType` creates a class that:

* Overrides all of the default implementations provided by ObjectIO
* Incorrectly documents the methods that it overrides
* Calls methods in the Core Type that may not exist

**Comment by samreid** [comment]
> All of my Core Types are missing one or more of those required methods.

There are 3 different types of serialization.  https://github.com/phetsims/phet-io/blob/3865e3e116822ecd46a18646e999213ed77caf59/doc/phet-io-instrumentation-guide.md#three-types-of-deserialization, each uses a different combination of methods.  They are not all required.

> Uses inconsistent parameter names and incorrect type expressions for the methods that it overrides

ObjectIO has a parameter of type `{*}` since it supports `{string}` and `{number}` as well as `{PhetioObject}`.  `createIOType` has a parameter of type `{PhetioObject}` since it only works with arguments of type `{PhetioObject}`.

For dealing with `class AlleleIO extends ReferenceIO( ObjectIO ) {}` here are some options:
1. Don't use ObjectIO.createIOType (since we don't want to forward to Allele)
2. Implement `toStateObject` and `fromStateObject` in Allele.js--they can forward to ReferenceIO
3. Add a new class ReferencePhetioObject and change Allele.js to extend it instead of PhetioObject, and add those forwarding methods in ReferencePhetioObject.  If we do this, we would combine with the strategy of adding PhetioObject.toStateObject, PhetioObject.fromStateObject etc as well.

**Comment by pixelzoom** [comment]
> There are 3 different types of serialization. ... They are not all required.

Yes, I'm aware of that. That doesn't mean that it's OK to generate a class that will fail if one of those methods is called, and can't use the default implementation. 

> ObjectIO has a parameter of type {*} since it supports {string} and {number} as well as {PhetioObject}. createIOType has a parameter of type {PhetioObject} since it only works with arguments of type {PhetioObject}.

How does that explain the difference in return types or parameter names?

**Comment by pixelzoom** [comment]
- [ ] `@param {string} ioTypeName` should be named `ioTypeName`

**Comment by pixelzoom** [comment]
- [ ] undocumented options:

```
    events: [],
    parameterTypes: []
```

**Comment by zepumph** [comment]
What would be a better implementation for a state method in `createIOType`?

1. Assert existence of all methods:

```js
    // @public
    static fromStateObject( stateObject ) {
      assert && assert( typeof coreType.fromStateObject );
      return coreType.fromStateObject( stateObject );
    }
```

2. Be graceful and defer to supertype when coreType doesn't have the method.

```js
    // @public
    static fromStateObject( stateObject ) {
      options.parentIOType
      return coreType.fromStateObject? coreType.fromStateObject( stateObject ): options.parentIOType.fromStateObject(stateObject);
    }

**Comment by pixelzoom** [comment]
I've created my own createIOType.js in natural-selection to address problems and deficiencies with ObjectIO.createIOType.   As noted in https://github.com/phetsims/natural-selection/issues/218#issuecomment-686661186:

> I bailed on using ObjectIO.createIOType and wrote my own createIOType.js that handles the IO Type for all of this sims IO Types, including parameterized ReferenceIO subclasses.
> 
> I'll note in phetsims/tandem#188, and the iO team is welcome to borrow from this for ObjectIO.createIOType. If ObjectIO.createIOType is cleaned up and addressed all of the needs of this sim, I'll be happy to revisit using it. Until then, closing this issue.

**Comment by pixelzoom** [comment]
Some notes about usage of natural-selection's createIOType.js ...  

Here's an example that's typical of how I used it.  I'm basically still delegating to the Core Type. This is nice and compact. And I did it this way since BunnyIO.js was delegating to the Core Type, so I didn't have to change much code.

 ```js
Bunny.BunnyIO = createIOType( Bunny, 'BunnyIO', {
  toStateObject: bunny => bunny.toStateObject(),
  stateToArgsForConstructor: state => Bunny.stateToArgsForConstructor( state ),
  applyState: ( bunny, stateObject ) => bunny.applyState( stateObject )
} );
```

Instead of delegating to the Core Type, I could have put the full implementation of these functions in the options to `createIOType`, something like:

 ```js
Bunny.BunnyIO = createIOType( Bunny, 'BunnyIO', {

  toStateObject: bunny => {
    return {
      ...
    };
  },

  stateToArgsForConstructor: state => {
    return [...];
  },

  applyState: ( bunny, stateObject ) => {
    ...
  }
} );
```

Either approach seems fine to me. I have a slight preference for delegating in the cases where these functions access private members of the Core Type.

**Comment by samreid** [comment]
The proposed usage site is:

```js
// Strategy A
Bunny.BunnyIO = createIOType( Bunny, 'BunnyIO', {
 toStateObject: bunny => bunny.toStateObject(),
 stateToArgsForConstructor: state => Bunny.stateToArgsForConstructor( state ),
 applyState: ( bunny, stateObject ) => bunny.applyState( stateObject )
} );
```

Compare this with using a `class` declaration in the core file like so:

```js
// Strategy B
Bunny.BunnyIO = class BunnyIO extends ObjectIO {
  static toStateObject( bunny ) { return bunny.toStateObject();  }
  static stateToArgsForConstructor( state ) { return Bunny.stateToArgsForConstructor( state ); }
  static applyState( bunny, state ) { bunny.applyState( state ); }
};
```

Since we are taking explicit control over the delegation, it seems preferable to define our own `class ... extends ...` to address the concerns raised in https://github.com/phetsims/tandem/issues/188#issuecomment-684167453.  This will also be much clearer than the dynamic dispatch happening in the `createIOType` implementations that dynamically decide whether to delegate or call to `super`.

```js
options.applyState ? options.applyState( object, state ) : super.applyState( object, state );
```

Like Strategy A, Strategy B has the advantage over the status quo that `Bunny` is defined in the same file, and could be validated like so:
```js
Bunny.BunnyIO.validator = { valueType:Bunny };
```

The main point of creating `createIOType` was to automatically delegate to the core type methods.  However, are we realizing that due to IO Type inheritance, etc, that we prefer to manually write these delegation functions on a case-by-case basis anyways?

More about how the implementation details of natural-selection/js/common/model/createIOType.js could be retained if we go to Strategy B:

### Documentation.
`createIOType` defines ``IO Type for ${ioTypeName.substring( 0, ioTypeName.length - IO_TYPE_SUFFIX.length )}``.  This could be defaulted in ObjectIO-- if it has a null `documentation`, it could use the typeName to infer the documentation as above.

### Validation
`createIOType` calls `validate( object, this.validator );` at the beginning of `toStateObject` and `applyState`.  We should omit these and just validate before calls to `toStateObject` and `applyState` in the PhetioStateEngine.  Or in the worst case scenario, we could streamline the syntax for that and include it in each call site:

```js
  static toStateObject( bunny ) { return validate(bunny) && bunny.toStateObject();  }
  static stateToArgsForConstructor( state ) { return Bunny.stateToArgsForConstructor( state ); }
  static applyState( bunny, state ) { validate(bunny) && bunny.applyState( state ); }
```

### Summary
In other cases where we introduced our own patterns for type creation (in `inherit` and `phetioInherit` and even mixins, we have incurred technical debt because it is not supported by the language.  By moving to use `class extends` in the idiomatic way, things work as expected and it's clear what's happening.  If we are not "autodelegating" to the core type, then I recommend we stick with `class BunnyIO extends ObjectIO`.  Any other benefits conferred by `createObjectIO` should be moved to the `class ... extends` strategy as much as possible.

### Plot twist
Somewhat contradicting the last paragraph (more like going to a medium point on the spectrum)... Could we keep the benefits of using `class ... extends` but still delegate to the core type like so?

```js
Bunny.BunnyIO = class BunnyIO extends ObjectIO {};

// Just delegate for these 3 methods
delegate( Bunny.BunnyIO, Bunny, [ 'toStateObject', 'stateToArgsForConstructor', 'applyState' ] );
```

Or it could detect what methods are implemented in the core type and auto-forward those:

```js
Bunny.BunnyIO = class BunnyIO extends ObjectIO {};

// Delegate all defined methods
delegate( Bunny.BunnyIO, Bunny );
```

### Summary Part II
After brainstorming the previous `delegate` pattern , I'm not sure it's worth the complexity.  After all, it is simple and straightforward to write forwarding methods like:

```js
Bunny.BunnyIO = class BunnyIO extends ObjectIO {
  static toStateObject( bunny ) { bunny.toStateObject();  }
};
```

Is it boilerplate? Yes. Is it boilerplate that is likely to be buggy and require a lot of maintenance?  Probably not.

Let's discuss, and consider that a short zoom call may save a lot of back and forth here in the issue.

**Comment by pixelzoom** [comment]
Happy to have a Zoom call, but I'm out until Tuesday 9/8.

The main thing that Strategy A is currently buying us over Strategy B is filling in this boilerplate:

```js
  IOType.documentation = options.documentation;
  IOType.validator = { valueType: coreType };
  IOType.typeName = ioTypeName;
  IOType.events = options.events;
  IOType.parameterTypes = options.parameterTypes;
  IOType.methods = options.methods;
  ObjectIO.validateSubtype( IOType );
```

Otherwise I'd be defining a class explicitly (e.g `class BunnyIO extends ObjectIO`), since it's absolutely more straightforward and readable.  The `class` approach still supports delegation, the boilerplate to do that is small/tolerable, and it's unlikely that I would use some like `delegate.`

**Comment by pixelzoom** [comment]
Here's a working patch that combines Strategy A and B.   The class is defined explicitly using `class extends`.  The boilerplate of static fields is filled in by `defineIOTypeFields`. BunnyIO definition looks like:

```js
Bunny.BunnyIO = class BunnyIO extends ObjectIO {

  static toStateObject( bunny ) { return bunny.toStateObject(); }

  static stateToArgsForConstructor( state ) { return Bunny.stateToArgsForConstructor( state ); }

  static applyState( bunny, stateObject ) { bunny.applyState( stateObject ); }
}
defineIOTypeFields( Bunny.BunnyIO, Bunny, 'BunnyIO' );
```

<details>
<summary>patch</summary>

```diff
Index: js/common/model/defineIOTypeField.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/defineIOTypeField.js	(date 1599227459959)
+++ js/common/model/defineIOTypeField.js	(date 1599227459959)
@@ -0,0 +1,47 @@
+// Copyright 2020, University of Colorado Boulder
+
+/**
+ * defineIOTypeFields handles the boilerplate of filling in defining static fields for an IO Type.
+ *
+ * @author Chris Malley (PixelZoom, Inc.)
+ */
+
+import merge from '../../../../phet-core/js/merge.js';
+import ObjectIO from '../../../../tandem/js/types/ObjectIO.js';
+import naturalSelection from '../../naturalSelection.js';
+
+// constants
+const IO_TYPE_SUFFIX = 'IO';
+
+function defineIOTypeFields( IOType, coreType, ioTypeName, options ) {
+
+  options = merge( {
+
+    // {string} see ObjectIO.documentation. The default is (e.g.) 'IO Type for Bunny'
+    documentation: `IO Type for ${ioTypeName.substring( 0, ioTypeName.length - IO_TYPE_SUFFIX.length )}`,
+
+    // see ObjectIO.methods
+    methods: {},
+
+    // see ObjectIO.events
+    events: [],
+
+    // see ObjectIO.parameterTypes
+    parameterTypes: []
+
+  }, options );
+
+  // Fill in static fields in the IO Type.
+  IOType.validator = { valueType: coreType };
+  IOType.typeName = ioTypeName;
+  IOType.documentation = options.documentation;
+  IOType.methods = options.methods;
+  IOType.events = options.events;
+  IOType.parameterTypes = options.parameterTypes;
+
+  // Verify that we've defined a valid IO Type.
+  ObjectIO.validateSubtype( IOType );
+}
+
+naturalSelection.register( 'defineIOTypeFields', defineIOTypeFields );
+export default defineIOTypeFields;
\ No newline at end of file
Index: js/common/model/Bunny.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/Bunny.js	(revision 58da6597272e2b2b084a90229e2b1ece97bcee7d)
+++ js/common/model/Bunny.js	(date 1599227796371)
@@ -19,10 +19,11 @@
 import BooleanIO from '../../../../tandem/js/types/BooleanIO.js';
 import NullableIO from '../../../../tandem/js/types/NullableIO.js';
 import NumberIO from '../../../../tandem/js/types/NumberIO.js';
+import ObjectIO from '../../../../tandem/js/types/ObjectIO.js';
 import ReferenceIO from '../../../../tandem/js/types/ReferenceIO.js';
 import naturalSelection from '../../naturalSelection.js';
 import NaturalSelectionUtils from '../NaturalSelectionUtils.js';
-import createIOType from './createIOType.js';
+import defineIOTypeFields from './defineIOTypeField.js';
 import EnvironmentModelViewTransform from './EnvironmentModelViewTransform.js';
 import GenePool from './GenePool.js';
 import Genotype from './Genotype.js';
@@ -441,11 +442,14 @@
  * as described in the Serialization section of
  * https://github.com/phetsims/phet-io/blob/master/doc/phet-io-instrumentation-guide.md#serialization
  */
-Bunny.BunnyIO = createIOType( Bunny, 'BunnyIO', {
-  toStateObject: bunny => bunny.toStateObject(),
-  stateToArgsForConstructor: state => Bunny.stateToArgsForConstructor( state ),
-  applyState: ( bunny, stateObject ) => bunny.applyState( stateObject )
-} );
+Bunny.BunnyIO = class BunnyIO extends ObjectIO {
+  static toStateObject( bunny ) { return bunny.toStateObject(); }
+
+  static stateToArgsForConstructor( state ) { return Bunny.stateToArgsForConstructor( state ); }
+
+  static applyState( bunny, stateObject ) { bunny.applyState( stateObject ); }
+}
+defineIOTypeFields( Bunny.BunnyIO, Bunny, 'BunnyIO' );
 
 naturalSelection.register( 'Bunny', Bunny );
 export default Bunny;
\ No newline at end of file
```

</details>

What we've lost here is:
* `validate` calls in `toStateObject` and `applyState`, could just add these in the method overrides
* `defineIOTypeFields` is also calling `ObjectIO.validateSubtype`, which should not be its responsibility, could just add this call after calling `defineIOTypeFields`

**Comment by pixelzoom** [comment]
For reference serialization, this looks nice:

```js
Allele.AlleleIO = class AlleleIO extends ReferenceIO( ObjectIO ){}
defineIOTypeFields( AlleleIO, Allele, 'AlleleIO' );
```

**Comment by samreid** [comment]
This seems promising, but I want to talk about the difference between:

```js
defineIOTypeFields( AlleleIO, Allele, 'AlleleIO',{

   // Strategy A
   parameterTypes:[X,Y,Z]
} );

// Strategy B
AlleleIO.parameterTypes = [X,Y,Z];
```

I like that Strategy A is declarative instead of imperative, but strategy B seems to be a better idiomatic fit for static class attribute declarations, and will be amenable to navigation, highlight, code complete, etc.

So I think we can keep some parts of `defineIOTypeFields` but several parts can move out like Strategy B.

**Comment by pixelzoom** [comment]
Seems like we're going to be left with:

```js
function defineIOTypeFields( IOType, coreType, ioTypeName ) {
  IOType.validator = { valueType: coreType };
  IOType.typeName = ioTypeName;
  IOType.documentation = `IO Type for ${ioTypeName.substring( 0, ioTypeName.length - IO_TYPE_SUFFIX.length )}`;
}
```

Then is it worth the bother, or should we just live with the boilerplate?

**Comment by pixelzoom** [comment]
In the above commit, I renamed `defineIOTypeFields` to `setIOTypeFields` and converted NS to this approach, deleting createIOType.js.   This felt like a step in the right direction.  Let's discuss.

**Comment by samreid** [comment]
That proposal looks good to me.  Here's a minor change that eliminates one temporary variable and shaves off 2 lines.

<details>

```diff
Index: js/common/model/Bunny.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/model/Bunny.js	(revision c33af929126426f6058bdb53137527551a68a5da)
+++ js/common/model/Bunny.js	(date 1599319103578)
@@ -441,8 +441,9 @@
  * it delegates to Bunny. The methods that BunnyIO overrides are typical of 'Dynamic element serialization',
  * as described in the Serialization section of
  * https://github.com/phetsims/phet-io/blob/master/doc/phet-io-instrumentation-guide.md#serialization
+ * @public
  */
-class BunnyIO extends ObjectIO {
+Bunny.BunnyIO = class extends ObjectIO {
 
   // @public @override
   static toStateObject( bunny ) { return bunny.toStateObject(); }
@@ -452,12 +453,9 @@
 
   // @public @override
   static applyState( bunny, stateObject ) { bunny.applyState( stateObject ); }
-}
+};
 
-setIOTypeFields( BunnyIO, 'BunnyIO', Bunny );
-
-// @public
-Bunny.BunnyIO = BunnyIO;
+setIOTypeFields( Bunny.BunnyIO, 'BunnyIO', Bunny );
 
 naturalSelection.register( 'Bunny', Bunny );
 export default Bunny;
\ No newline at end of file

```

</details>

I feel like this is a good direction to go, but I don't want to start converting all repos to use this pattern until @zepumph reviews and comments.

**Comment by pixelzoom** [comment]
I prefer this pattern:

```
class CoreIOType extends SuperIOType {...}
setIOTypeFields(...)
CoreType.CoreIOType = CoreIOType;
```

... because the class is complete and validated before assigning to a static constant.  But I don't mind if others use the pattern that you've suggested in the patch.

**Comment by pixelzoom** [comment]
> UPDATE: One more review comment, I think setIOTypeFields should take an options argument in case a different documentation needs to be specified.

OK with me. But just pointing out that doc could also be added by explicit assignment:

```js
CoreType.CoreIOType = class CoreIOType extends SuperIOType {...}
setIOTypeFields( CoreIOType, 'CoreIOType', CoreType )
CoreIOType.documentation = '....';
```

How do you feel about `setIOTypeFields` calling `ObjectIO.validateSubtype( IOType )`?

**Comment by pixelzoom** [comment]
Also might be good to rename ObjectIO `validateSubtype` to `validateIOType`.  There's no reason it can't validate any IO Type, including itself, right?

**Comment by samreid** [comment]
> But just pointing out that doc could also be added by explicit assignment:

I agree and that wouldn't be too bad, but then there would be a period of time where the IO Type has the wrong value, and it may be confusing that the value gets assigned twice.

> How do you feel about `setIOTypeFields` calling `ObjectIO.validateSubtype( IOType )`?

I agree the validation belongs in that method.  I considered whether `setIOTypeFields` should have a different name, like `initializeIOType` since it is doing more than just setting IO Type fields.  What do you think?

> Also might be good to rename ObjectIO `validateSubtype` to `validateIOType`.  There's no reason it can't validate any IO Type, including itself, right?  

Yes, that would be a better name.  Indeed it already does validate itself.  Would you like to do that step, or let me know if I should.

**Comment by samreid** [comment]
Also, how much of this will we want to revisit once we can use static properties, like so?

```js
class BunnyIO extends ObjectIO {
  static documentation = '...';
}
```

**Comment by pixelzoom** [comment]
ObjectIO.validateSubtype was renamed validateIOType in the above commits.

**Comment by pixelzoom** [comment]
options.documentation was added to `setIOTypeFields` in https://github.com/phetsims/natural-selection/commit/5a09377d000076285c2733edf59619c77a426828.

Should `setIOTypeFields` be moved to ObjectIO?  And should `createIOType` be deleted from ObjectIO?

**Comment by pixelzoom** [comment]
Also, how much of this will we want to revisit once we can use static properties, like so?

```js
class BunnyIO extends ObjectIO {
  static documentation = '...';
}
```

Yes, undoubtedly.  It's the penalty for using a wannabe language that can't seem to get OO right.

**Comment by samreid** [comment]
According to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields there is already support for static class fields on all our supported platforms.  There is also a babel plugin which will transform that correctly: https://babeljs.io/docs/en/babel-plugin-proposal-class-properties.

> But that still won't give us a reasonable default for `documentation`.

First, a reasonable default for documentation could be `null` or `''` and then in Studio or other places that display the IO Type documentation, it could patch it lazily, like `display(ioType.documentation || `IO Type for ${ioType.typeName.substring(...)}`. 

Second, we do need to call `ObjectIO.validateIOType( ioType )` on each IO Type. The documentation "patching" could happen there as well, or we could rename it something more general like `initializeIOType`.

Sketching out what this could look like: 

```js

// Strategy A
class BunnyIO extends ObjectIO {
  static toStateObject( bunny ) { return bunny.toStateObject(); } // @public @override
  static stateToArgsForConstructor( state ) { return Bunny.stateToArgsForConstructor( state ); } // @public @override
  static applyState( bunny, stateObject ) { bunny.applyState( stateObject ); } // @public @override
  static documentation = 'My Documentation';
  static typeName = 'BunnyIO';
  static validator = { valueType: Bunny };
}

initializeIOType( BunnyIO );
```

Comparing to the previous proposal:
```js
// Strategy B
class BunnyIO extends ObjectIO {
  static toStateObject( bunny ) { return bunny.toStateObject(); } // @public @override
  static stateToArgsForConstructor( state ) { return Bunny.stateToArgsForConstructor( state ); } // @public @override
  static applyState( bunny, stateObject ) { bunny.applyState( stateObject ); } // @public @override
}

initializeIOType( BunnyIO,'BunnyIO',Bunny,{
  documentation: 'My documentation'
} );
```

**Comment by pixelzoom** [comment]
FYI...  NS milestone is 9/17 to begin QA testing on a dev version (brand=phet-io) for delivery to the client.  After that we'll be jumping right into a full QA process for production.  This issue is not blocking for any of that.  If not completed by 9/17, then we'll publish with the NS-specific implementation in setIOTypeField.js for https://github.com/phetsims/natural-selection/issues/218.

**Comment by pixelzoom** [comment]
9/9/2020 Zoom meeting with @samreid and @zepumph:

* For the near term (and especially the 10/1/2020 NS deliverable) we'll go with what's currently being done in NS setIOTypeFields.js.   I will move `setIOTypeFields` to a static method of ObjectIO.  @samreid or @zepumph will delete `createIOType` and its usages.

* For the longterm, there are other possible IO Type patterns to explore.  Until shas have been taken for 10/1/2020 deliverable, exploration will be done in branches or patches.  (shas should be taken around 9/17/2020).

**Comment by pixelzoom** [comment]
setIOTypeFields.js has been moved to `ObjectIO.setIOTypeFields`, and is now used by NS.  Self unassigning until there's more to do here.

**Comment by samreid** [comment]
I reverted usages of `createIOType` and deleted it.  This should block publication until those changes have been reviewed.  I'd like to request a review for the changes in Action.js and Emitter.js by @zepumph before NS SHAs.  If you don't have time, let's ask @pixelzoom if he can review this. Once that is complete, this issue can be closed.  Side tasks will be moved to new issues as described below.

Summarizing results from this issue:

* Automatically forwarding to a core type doesn't work well, because it is too tricky and is confusing regarding when to call `IOType.superMethod` vs `CoreType.method`.  Manually written delegation methods are clear and succinct.
* We prefer to move the IO types to the bottom of the core type file.  This makes it possible to set the `validator` using the exact class without needing to access it lazily or access it from the namespace.  I'll make a new issue to move to this pattern.
* We need to investigate improved ways to describe IO types, how to describe an implement an interface.  This will be pursued in https://github.com/phetsims/tandem/issues/211
* We can use `setIOTypeFields` for now, but ultimately may abandon that once we can declare `static` class attributes.  The direction in https://github.com/phetsims/tandem/issues/211 will determine how we proceed on this aspect.
* The delegation methods are meant to be succinct, and do not warrant elaborate JSDoc.  Previously we used to have:

```js
class BunnyIO extends ObjectIO {

  /**
   * Serializes a Bunny instance.
   * @param {Bunny} bunny
   * @returns {Object}
   * @public
   * @override
   */
  static toStateObject( bunny ) {
    validate( bunny, this.validator );
    return bunny.toStateObject();
  }

  /**
   * Creates the args to BunnyGroup.createNextElement that creates Bunny instances.
   * @param state
   * @returns {Object[]}
   * @public
   * @override
   */
  static stateToArgsForConstructor( state ) {
    return Bunny.stateToArgsForConstructor( state );
  }

  /**
   * Restores Bunny state after instantiation.
   * @param {Bunny} bunny
   * @param {Object} state
   * @public
   * @override
   */
  static applyState( bunny, state ) {
    validate( bunny, this.validator );
    bunny.applyState( state );
  }
}
```

This has been compressed to:

```js
class BunnyIO extends ObjectIO {

  // @public @override
  static toStateObject( bunny ) { return bunny.toStateObject(); }

  // @public @override
  static stateToArgsForConstructor( state ) { return Bunny.stateToArgsForConstructor( state ); }

  // @public @override
  static applyState( bunny, stateObject ) { bunny.applyState( stateObject ); }
}
```

**Comment by pixelzoom** [comment]
@samreid said:
> Previously we used to have:
> ...
> This has been compressed to:

There's not any real compression here. The compression is due to 3 question things in the example:

* omission of `validate` call, likely may not be an improvement
* abbreviation of method doc and omission of `@param` annotation, again likely not an improvement
* defining what's left of the method on 1 line

So the code is essentially identical, and there's really no difference here. 

The compression was actually achieved in definition of static fields. For example this:

```js
BunnyIO.documentation = 'IO Type for Bunny';
 BunnyIO.validator = { isValidValue: value => value instanceof Bunny };
 BunnyIO.typeName = 'BunnyIO';
 ObjectIO.validateSubtype( BunnyIO );
```

was replaced with:

```js
ObjectIO.setIOTypeFields( BunnyIO, 'BunnyIO', Bunny );
```

**Comment by pixelzoom** [comment]
Otherwise the IO Types were just moved to same file as their associated Core Type. And they could just as easily have been left in their own .js files. It's really only necessary to move the IO Type definition into the Core Type .js file if you need to avoid circular references between IO Type and Core Type.

**Comment by samreid** [comment]
I've written a proposal in https://github.com/phetsims/tandem/issues/211#issuecomment-691240527 which I think we should pursue.  It automatically includes the `validate` calls, eliminates the need for JSDoc, and eliminates the need for `setIOTypeFields` and `validateIOType`.  It's a larger change which will require changes to all IO Types and review, and we can hold off on it until Natural Selection has safe SHAs.

**Comment by zepumph** [comment]
Given the comment over in https://github.com/phetsims/tandem/issues/211#issuecomment-692983485, this doesn't seem to block publication. @samreid, I'm going to remove this from that list, please change if there is another reason not to. Does this block NS?

**Comment by pixelzoom** [comment]
This issue is definitely not blocking for the 10/1 RC deliverable, the subject of https://github.com/phetsims/natural-selection/issues/208.  As soon as that RC has been delivered, then this issue will immediately becoming blocking for the next NS milestone.

**Comment by samreid** [comment]
Let's put this on hold pending work in #211.

**Comment by samreid** [comment]
Some brainstorming in #211 revolved around automatically leveraging `IOType.valueType` if it exists.

**Comment by samreid** [comment]
The latest proposal for this step is in https://github.com/phetsims/tandem/issues/213, let's close this and continue there.
